File: src/main/java/no/mesan/reactive/actors/BildeAnsvarlig.java
package no.mesan.reactive.actors;

import java.util.HashSet;
import java.util.Set;

import no.mesan.reactive.actors.messages.FoundImage;
import no.mesan.reactive.actors.messages.ImageRequest;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.Props;
import akka.japi.pf.ReceiveBuilder;

public class BildeAnsvarlig extends AbstractActor {

    private final ActorRef klientAnsvarlig;
    private Set<String> foundImages = new HashSet<>();

    public BildeAnsvarlig(final ActorRef klientAnsvarlig) {
        this.klientAnsvarlig = klientAnsvarlig;

        receive(ReceiveBuilder
                        .match(FoundImage.class, this::handleFoundImage)
                        .matchAny(this::unhandled)
                        .build()
        );
    }

    private void handleFoundImage(final FoundImage foundImage) {
        final String imageUrl = foundImage.getImageUrl();
        if (isANewImage(imageUrl)) {
            foundImages.add(imageUrl);
            final ActorRef bildeLager = context().actorOf(Props.create(BildeLager.class, klientAnsvarlig));
            bildeLager.tell(new ImageRequest(foundImage.getImageUrl()), context().self());
        }
    }

    private boolean isANewImage(final String imageUrl) {
        return !foundImages.contains(imageUrl);
    }
}

File: src/main/scala/no/mesan/fag/reactive/scala/actors/BildeAnsvarlig.scala
package no.mesan.fag.reactive.scala.actors

import akka.actor.{ActorLogging, Props, ActorRef, Actor}

/** Holder orden på bildene vi har funnet. Lager nye thumbnails ved behov. */
class BildeAnsvarlig(klientAnsvarlig: ActorRef) extends Actor with ActorLogging with Tracer {
  import BildeAnsvarlig.FoundImage
  traceCreate(s"Bildeansvarlig for $klientAnsvarlig")

  def withFoundImages(foundImages: Set[String]): Receive = {
    case found: FoundImage =>
      def isANewImage(url: String) = !(foundImages contains url)
      if (isANewImage(found.url)) {
        found.traceReceive(this)
        val reader = context.actorOf(BildeLager.props(klientAnsvarlig))
        reader ! found
      }
      else
        found.traceReceive("ignored")
      context become withFoundImages(foundImages + found.url)
  }

  override def receive = withFoundImages(Set.empty)
}

/** Opprettelse og meldinger. */
object BildeAnsvarlig {
  /** Melding om et bilde som skal scannes. */
  case class FoundImage(url: String) extends ApplicationMessage

  /** Props for ny forekomst. */
  def props(klientAnsvarlig: ActorRef) = Props(classOf[BildeAnsvarlig], klientAnsvarlig)
}
